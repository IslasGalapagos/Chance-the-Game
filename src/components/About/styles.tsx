/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright © 2018 Evgeny Sysoletin. All rights reserved.
*/

import styled, {css} from 'react-emotion';

interface StyledWrapperProps {
  isShown: boolean;
}

export const StyledWrapper = styled('div')<StyledWrapperProps>`
  position: absolute;
  bottom: 0;
  right: 0;
  z-index: 3;

  ${({isShown}) =>
    isShown
      ? `
    width: 100%;
    height: 100%;

    background-color: rgba(219, 249, 255, 0.2);
  `
      : ''};
`;

export const textBlockCSS = css`
  position: absolute;
  bottom: 0;
  right: 0;
  z-index: 0;

  width: 50%;
  height: 50%;
  padding: 25px;
  box-sizing: border-box;

  background-color: #84b7b1;

  .about__text,
  .about__source_link,
  .about__copyright {
    color: #fff;
  }

  .about__text {
    font-size: 24px;

    margin: 0 0 15px 0;
  }

  .about__text,
  .about__copyright {
    font-family: 'Helvetica Neu', Helvetica, Arial;
  }

  .about__source_link {
    font-family: 'Courier New', Courer;
    font-size: 20px;
    font-weight: bold;
    text-decoration: none;

    display: inline-block;

    position: relative;

    margin: 15px 0 0 26px;

    border-bottom: 1px solid #fff;

    &:before {
      content: '⇨';

      position: absolute;
      top: -2px;
      left: -26px;
    }

    &:hover {
      opacity: 0.75;
    }
  }

  .about__copyright {
    font-size: 14px;

    position: absolute;
    left: 25px;
    bottom: 25px;
  }
`;

export const togglerCSS = css`
  position: absolute;
  right: 25px;
  bottom: 25px;
  z-index: 1;

  width: 74px;
  height: 74px;
  padding: 0;

  background-color: transparent;
  border: none;
  outline: none;

  cursor: pointer;

  &:before {
    content: '';

    position: absolute;
    left: 10px;
    top: 10px;

    width: 54px;
    height: 54px;
    box-sizing: border-box;

    border: 2px solid #c8e8e5;

    transform: rotate(45deg);
  }
`;
