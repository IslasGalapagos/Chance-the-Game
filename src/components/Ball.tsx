/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import * as React from 'react';
import StyledBall from './Ball.styles';
import {Positions} from './Game';

import '../static/ball-good.svg';
import '../static/ball-bad.svg';

export interface Props {
  chosenPosition: Positions;
  isWin: boolean;
}

class Ball extends React.PureComponent<Props, {}> {
  render() {
    const {isWin, chosenPosition} = this.props;

    return (
      <StyledBall
        src={`./images/ball-${isWin ? 'good' : 'bad'}.svg`}
        isWin={isWin}
        chosenPosition={chosenPosition}
        alt="ball"
      />
    );
  }
}

export default Ball;
