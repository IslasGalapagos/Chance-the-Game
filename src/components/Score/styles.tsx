/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import * as React from 'react';
import styled, {css} from 'react-emotion';

const NumCSS = css`
  font-family: 'Helvetica Neu', Helvetica, Arial;
  font-size: 36px;
  text-align: center;
  position: relative;
  display: inline-block;
  width: 64px;
  box-sizing: border-box;
  border-radius: 50%;

  transition: left 0.2s ease;
`;

interface NumProps {
  className?: string;
  score: number;
}

const Num: React.SFC<NumProps> = ({className, score}) => (
  <div className={`${NumCSS} ${className}`}>{score}</div>
);

interface StyledNumProps {
  user?: boolean;
  gameState: string;
}

export const StyledNum = styled(Num)<StyledNumProps>`
  line-height: ${({user}) => (user ? '60px' : '64px')};

  left: ${({gameState, user}) => {
    if (gameState === 'draw') {
      return user ? '1px' : '-1px';
    }

    return user ? '5px' : '-5px';
  }};
  z-index: ${({gameState, user}) => {
    if (gameState === 'draw') {
      return '';
    }

    if (user) {
      return gameState === 'user_lead' ? '1' : '0';
    }

    return gameState === 'random_lead' ? '1' : '0';
  }};

  background-color: ${({user}) => (user ? '#FCFFDA' : '#49404E')};
  color: ${({user}) => (user ? '#49404E' : '#FCFFDA')};
  border: ${({user}) => (user ? '2px solid #49404E' : '')};
`;

export const StyledWrapper = styled('div')`
  position: absolute;
  top: 35px;
  left: 35px;
  z-index: 1;
`;
