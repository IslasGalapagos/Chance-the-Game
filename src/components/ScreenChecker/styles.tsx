/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import styled from 'react-emotion';

const StyledStub = styled('div')`
  position: absolute;
  left: 0;
  top: 0;
  z-index: 5;

  display: flex;
  justify-content: center;
  align-items: center;

  width: 100%;
  height: 100%;

  background: #0000b1;

  span {
    font-family: 'Courier New', Courer;
    font-size: 25px;
    font-weight: bold;
    text-align: center;

    padding: 0 30px;

    color: #fff;

    @media (min-width: 740px) {
      padding: 0;
    }
  }
`;

export default StyledStub;
