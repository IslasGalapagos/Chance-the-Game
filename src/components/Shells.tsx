/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright © 2018 Evgeny Sysoletin. All rights reserved.
*/

import * as React from 'react';
import {StyledWrapper, StyledShell} from './Shells.styles';
import Ball from './Ball';
import {Positions} from './Game';

import '../static/shell-1.svg';
import '../static/shell-2.svg';
import '../static/shell-3.svg';

export interface Props {
  move: (position: Positions) => void;
  onChoose: (position: Positions) => void;
  onMouseLeave: () => void;
  chosenPosition: Positions | null;
  isWin: boolean | null;
}

class Shells extends React.PureComponent<Props, {}> {
  constructor(props: any) {
    super(props);

    this.onChoose = this.onChoose.bind(this);
  }

  onChoose(position: Positions): () => void {
    return () => {
      this.props.onChoose(position);
    };
  }

  onMouseEnter(position: Positions): () => void {
    return () => {
      this.props.move(position);
    };
  }

  render() {
    const {onMouseLeave, chosenPosition, isWin} = this.props;

    const shells = [Positions.Left, Positions.Center, Positions.Right].map(
      (position, index) => (
        <StyledShell
          src={`./images/shell-${index + 1}.svg`}
          alt={`shell № ${index + 1}`}
          key={position}
          onClick={this.onChoose(position)}
          onMouseEnter={this.onMouseEnter(position)}
          onMouseLeave={onMouseLeave}
          isHidden={chosenPosition !== null && chosenPosition === position}
          middle={index === 1}
        />
      )
    );

    return (
      <StyledWrapper>
        {shells}
        {isWin !== null &&
          chosenPosition !== null && (
            <Ball isWin={isWin} chosenPosition={chosenPosition} />
          )}
      </StyledWrapper>
    );
  }
}

export default Shells;
