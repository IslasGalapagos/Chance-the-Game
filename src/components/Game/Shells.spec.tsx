/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import {shallow, mount} from 'enzyme';
import * as testRenderer from 'react-test-renderer';
import * as React from 'react';
import Shells, {Props} from './Shells';
import {StyledShell as Shell} from './Shells.styles';
import Ball from './Ball';
import {Positions} from '.';

const props: Props = {
  move: jest.fn(),
  onChoose: jest.fn(),
  onMouseLeave: jest.fn(),
  chosenPosition: null,
  isWin: null
};

describe('<Shells/>', () => {
  it('renders correctly', () => {
    const treeAndStyles = testRenderer.create(<Shells {...props} />).toJSON();
    expect(treeAndStyles).toMatchSnapshot();
  });

  // SHELL
  it('calls props.onChoose with determined position on click for each shell', () => {
    const spyOnChoose = jest.spyOn(Shells.prototype, 'onChoose');
    const component = mount(<Shells {...props} />);

    expect(spyOnChoose).toBeCalledTimes(3);
    expect(spyOnChoose.mock.calls[0][0]).toEqual(Positions.Left);
    expect(spyOnChoose.mock.calls[1][0]).toEqual(Positions.Center);
    expect(spyOnChoose.mock.calls[2][0]).toEqual(Positions.Right);

    const spyPropsOnChoose = jest.spyOn(props, 'onChoose');

    component.find('img').forEach(el => el.simulate('click'));

    expect(spyPropsOnChoose).toBeCalledTimes(3);
    expect(spyPropsOnChoose.mock.calls[0][0]).toEqual(Positions.Left);
    expect(spyPropsOnChoose.mock.calls[1][0]).toEqual(Positions.Center);
    expect(spyPropsOnChoose.mock.calls[2][0]).toEqual(Positions.Right);
  });

  it('calls props.move with determined position on mouseenter for each shell', () => {
    const spyOnMouseEnter = jest.spyOn(Shells.prototype, 'onMouseEnter');
    const component = mount(<Shells {...props} />);

    expect(spyOnMouseEnter).toBeCalledTimes(3);
    expect(spyOnMouseEnter.mock.calls[0][0]).toEqual(Positions.Left);
    expect(spyOnMouseEnter.mock.calls[1][0]).toEqual(Positions.Center);
    expect(spyOnMouseEnter.mock.calls[2][0]).toEqual(Positions.Right);

    const spyMove = jest.spyOn(props, 'move');

    component.find('img').forEach(el => el.simulate('mouseenter'));

    expect(spyMove).toBeCalledTimes(3);
    expect(spyMove.mock.calls[0][0]).toEqual(Positions.Left);
    expect(spyMove.mock.calls[1][0]).toEqual(Positions.Center);
    expect(spyMove.mock.calls[2][0]).toEqual(Positions.Right);
  });

  it('calls props.onMouseLeave on mouseleave for each shell', () => {
    const spyonMouseLeave = jest.spyOn(props, 'onMouseLeave');
    const component = mount(<Shells {...props} />);

    expect(spyonMouseLeave).toBeCalledTimes(0);

    component.find('img').forEach(el => el.simulate('mouseleave'));

    expect(spyonMouseLeave).toBeCalledTimes(3);
  });

  it('sets bool dependent on props.chosenPosition to Shell.isHidden', () => {
    const component = shallow(<Shells {...props} />);

    const hiddenShells = component
      .find(Shell)
      .filterWhere(el => el.prop('isHidden') === true);

    expect(hiddenShells.length).toEqual(0);

    component.setProps({chosenPosition: Positions.Left});
    const shells = component.find(Shell);
    expect(shells.at(0).prop('isHidden')).toEqual(true);
    expect(shells.at(1).prop('isHidden')).toEqual(false);
    expect(shells.at(2).prop('isHidden')).toEqual(false);
  });

  // BALL
  it('shows ball when props.isWin is not null & props.chosenPosition is not null', () => {
    const component = shallow(<Shells {...props} />);
    expect(component.find(Ball).length).toEqual(0);
    component.setProps({isWin: true, chosenPosition: Positions.Center});
    expect(component.find(Ball).length).toEqual(1);
  });

  it(`sets props.isWin to Ball's isWin`, () => {
    const component = shallow(
      <Shells {...props} isWin={true} chosenPosition={Positions.Center} />
    );
    expect(component.find(Ball).prop('isWin')).toEqual(true);
  });

  it(`sets props.chosenPosition to Ball's chosenPosition`, () => {
    const component = shallow(
      <Shells {...props} isWin={true} chosenPosition={Positions.Center} />
    );
    expect(component.find(Ball).prop('chosenPosition')).toEqual(
      Positions.Center
    );
  });
});
