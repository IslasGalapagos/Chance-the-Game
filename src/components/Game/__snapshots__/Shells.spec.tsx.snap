// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Shells/> renders correctly 1`] = `
.emotion-3 {
  position: relative;
  display: block;
}

.emotion-0 {
  display: inline-block;
  vertical-align: bottom;
  cursor: pointer;
}

.emotion-0:hover {
  -webkit-animation: animation-qym1sx 0.2s ease infinite;
  animation: animation-qym1sx 0.2s ease infinite;
  -webkit-animation-delay: 0.2s;
  animation-delay: 0.2s;
}

.emotion-1 {
  margin: 0 50px;
  display: inline-block;
  vertical-align: bottom;
  cursor: pointer;
}

.emotion-1:hover {
  -webkit-animation: animation-qym1sx 0.2s ease infinite;
  animation: animation-qym1sx 0.2s ease infinite;
  -webkit-animation-delay: 0.2s;
  animation-delay: 0.2s;
}

<div
  className="emotion-3"
>
  <img
    alt="shell № 1"
    className="emotion-0"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[MockFunction]}
    src="./images/shell-1.svg"
  />
  <img
    alt="shell № 2"
    className="emotion-1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[MockFunction]}
    src="./images/shell-2.svg"
  />
  <img
    alt="shell № 3"
    className="emotion-0"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[MockFunction]}
    src="./images/shell-3.svg"
  />
</div>
`;
